File: 알고리즘_도감/gnsals0904/week2.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;

class week2 {
    public static void main(String[] args) {
        System.out.println(solution1("([{}])"));
        System.out.println(solution1("(]"));
        System.out.println(solution1("hello (world)"));
        System.out.println(Arrays.toString(solution2(3, new int[] {97, 98, 99, 100, 101})));
        System.out.println(Arrays.toString(solution2(1, new int[] {10, 20, 30})));
        System.out.println(Arrays.toString(solution2(5, new int[] {1, 2})));
        System.out.println(solution3("This is a test. This test is good."));
        System.out.println(solution3(""));
        System.out.println(solution4(new int[] {4, 10, 3, 5, 1}, 3));
        System.out.println(solution5(new int[] {5, 3, 7, 2, 4, 6, 8}));
        System.out.println(solution5(new int[] {5, 3, 8, 3, 1}));
        System.out.println(solution5(new int[] {}));
    }

    /**
     * 1. 괄호 유효성 검사
     *   - stack으로 구현
     *   - 스택에서 나온 괄호가 정확히 내가 들고 있는 괄호와 짝이 맞아야함
     *   - 문자열 끝까지 검토했을때 내가 손에 들고 있는 괄호가 없어야 함.
     * @param input
     * @return
     */
    static boolean solution1(String input) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < input.length(); i++) {
            char current = input.charAt(i);

            if (current == '(' || current == '{' || current == '[') {
                stack.push(current);
            }

            if (current == ')' || current == '}' || current == ']') {
                if (stack.isEmpty()) {
                    return false;
                }

                char open = stack.pop();

                if (current == ')' && open != '(') {
                    return false;
                }

                if (current == '}' && open != '{') {
                    return false;
                }

                if (current == ']' && open != '[') {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    /**
     * 2. 최근 N개의 요청 추적기
     *   - 큐의 크기가 N보다 커지면 가장 오래된 요청을 뺌
     *   - 각 요청을 처리한 뒤 현재 큐의 크기를 반환 배열에 저장
     * @param n
     * @param requests
     * @return
     */
    static int[] solution2(int n, int[] requests) {
        Queue<Integer> queue = new LinkedList<>();
        int[] result = new int[requests.length];

        for (int i = 0; i < requests.length; i++) {
            queue.add(requests[i]);

            if (queue.size() > n) {
                queue.poll();
            }

            result[i] = queue.size();
        }

        return result;
    }

    /**
     * 3. 단어 빈도수 계산
     *   - 대소문자는 구분하지 않으므로 소문자로 변환
     *   - 구두점을 제거한 뒤 공백 기준으로 단어를 나눔
     *   - HashMap에 단어별 개수를 저장
     * @param input
     * @return
     */
    static Map<String, Integer> solution3(String input) {
        Map<String, Integer> result = new HashMap<>();
        String cleanedInput = input.toLowerCase().replaceAll("\\p{Punct}", "").trim();

        if (cleanedInput.isEmpty()) {
            return result;
        }

        String[] words = cleanedInput.split("\\s+");

        for (String word : words) {
            result.put(word, result.getOrDefault(word, 0) + 1);
        }

        return result;
    }

    /**
     * 4. K번째로 작은 값 찾기
     *   - PriorityQueue는 기본적으로 최소 힙
     *   - 모든 값을 넣은 뒤 k번 꺼내면 k번째로 작은 값
     * @param nums
     * @param k
     * @return
     */
    static int solution4(int[] nums, int k) {
        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>();

        for (int num : nums) {
            priorityQueue.add(num);
        }

        int result = 0;

        for (int i = 0; i < k; i++) {
            result = priorityQueue.poll();
        }

        return result;
    }

    /**
     * 5. 이진 탐색 트리 삽입
     *   - 입력 순서대로 BST에 삽입
     *   - 중복 값은 무시
     *   - 완성된 트리를 BFS로 순회
     *   - 마지막 레벨의 의미 없는 null은 제거
     * @param nums
     * @return
     */
    static List<Integer> solution5(int[] nums) {
        if (nums.length == 0) {
            return new ArrayList<>();
        }

        TreeNode root = new TreeNode(nums[0]);

        for (int i = 1; i < nums.length; i++) {
            insert(root, nums[i]);
        }

        List<Integer> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();

            if (current == null) {
                result.add(null);
            } else {
                result.add(current.val);
                queue.add(current.left);
                queue.add(current.right);
            }
        }

        while (!result.isEmpty() && result.get(result.size() - 1) == null) {
            result.remove(result.size() - 1);
        }

        return result;
    }

    static void insert(TreeNode node, int value) {
        if (value < node.val) {
            if (node.left == null) {
                node.left = new TreeNode(value);
            } else {
                insert(node.left, value);
            }
        }

        if (value > node.val) {
            if (node.right == null) {
                node.right = new TreeNode(value);
            } else {
                insert(node.right, value);
            }
        }
    }

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }
}

File: 알고리즘_도감/sgoldenbird/week2.js
// 1. 괄호 유효성 검사 
function solution1(S) {
  const stack = [];
  const pairs = {
    ')': '(',
    '}': '{',
    ']': '['
  };

  for (let char of S) {
    if (char === '(' || char === '{' || char === '[') {
      stack.push(char);
    } 
    else if (pairs[char]) {
      const topElement = stack.pop();
      if (topElement !== pairs[char]) {
        return false;
      }
    }
  }

  return stack.length === 0;
}
 

// 2. 최근 N개의 요청 추적기
function solution2(N, arr) {
  const queue = [];
  const count = [];
  
  for (let i = 0; i < arr.length; i++) {
    queue.push(arr[i]);
    
    if (queue.length > N) {
      // queue.shift();
      for (let j = 0; j < queue.length - 1; j++) {
        queue[j] = queue[j + 1];
      }
      queue.pop();
    }

    count[i] = queue.length;
}
  return count;
}

// 3. 단어 빈도수 계산
function solution3(sentence) {
  if (!sentence || sentence.trim() === "") {
    return {};
  }

  const lowerSentence = sentence.toLowerCase();
  const result = {};
  let currentWord = "";
  
  for (let i = 0; i < lowerSentence.length; i++) {
    const char = lowerSentence[i];

    if (char >= 'a' && char <= 'z') {
      currentWord += char;
    } else {
      if (currentWord.length > 0) {
        result[currentWord] = (result[currentWord] || 0) + 1;
        currentWord = ""; 
      }
    }
  }

  // 문장이 단어로 끝날 경우, 마지막 단어 처리
  if (currentWord.length > 0) {
    result[currentWord] = (result[currentWord] || 0) + 1;
  }

  return result;

}


// 4. K번째로 작은 값 찾기
class Heap {
  constructor() {
    this.heap = [];
  }

  push(val) {
    this.heap.push(val);
    this.bubbleUp();
  }

  pop() {
    if (this.heap.length === 0) return null;
    if (this.heap.length === 1) return this.heap.pop();

    const min = this.heap[0];
    this.heap[0] = this.heap.pop(); // 마지막 요소를 루트로 이동
    this.bubbleDown();
    return min;
  }

  bubbleUp() {
    let index = this.heap.length - 1;
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.heap[parentIndex] <= this.heap[index]) break;
      
      // 부모보다 작으면 교체
      [this.heap[parentIndex], this.heap[index]] = [this.heap[index], this.heap[parentIndex]];
      index = parentIndex;
    }
  }

  bubbleDown() {
    let index = 0;
    const length = this.heap.length;

    while (true) {
      let leftChildIndex = index * 2 + 1;
      let rightChildIndex = index * 2 + 2;
      let smallest = index;

      if (leftChildIndex < length && this.heap[leftChildIndex] < this.heap[smallest]) {
        smallest = leftChildIndex;
      }

      if (rightChildIndex < length && this.heap[rightChildIndex] < this.heap[smallest]) {
        smallest = rightChildIndex;
      }

      if (smallest === index) break;

      [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
      index = smallest;
    }
  }
}

function solution4(arr, K) {
  
  let result;
  const heap = new Heap();

 
  for (const num of arr) {
    heap.push(num);
  }

  for (let i = 0; i < K; i++) {
    result = heap.pop();
  }

  return result;

}

/* 5. 이진 탐색 트리 삽입
- 1단계: 배열의 숫자를 하나씩 꺼내 이진 탐색 트리(BST)를 만든다. 
- 2단계: 만들어진 트리를 층별로 훑으면서(BFS) 배열로 다시 만든다. 
*/
class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

function solution5(arr) {
  if (arr.length === 0) return [];

  let root = null;

  // 1단계: BST 삽입 로직
  const insert = (node, val) => {
    if (node === null) return new TreeNode(val);

    if (val < node.val) {
      node.left = insert(node.left, val);
    } else if (val > node.val) {
      node.right = insert(node.right, val);
    }
    // 값이 같으면(중복) 아무것도 하지 않고 현재 노드 반환
    return node;
  };

  // 모든 숫자를 트리에 삽입
  arr.forEach((num) => {
    root = insert(root, num);
  });

  // 2단계: BFS(레벨 순회) 로직
  const result = [];
  const queue = [root];

  while (queue.length > 0) {
    const current = queue.shift();

    if (current) {
      result.push(current.val);
      // 자식이 null이더라도 일단 큐에 넣어서 위치를 표시함
      queue.push(current.left);
      queue.push(current.right);
    } else {
      result.push(null);
    }
  }

  // 3단계: 마지막의 trailing null 제거
  while (result.length > 0 && result[result.length - 1] === null) {
    result.pop();
  }

  return result;

}